File: backend/services/session.go
package services

import (
	"backend/claims"
	"backend/database"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/jackc/pgx/v5"
)

const (
	findSessionTimeout   = time.Second * 5
	revokeSessionTimeout = time.Second * 5
)

var (
	ErrSessionNotFound = fiber.NewError(404, "session not found")
)

type FindSessionRequest struct {
	Offset      int  `json:"offset" validate:"min=0"`
	Count       int  `json:"count" validate:"required,min=1,max=100"`
	ShowRevoked bool `json:"showRevoked"`
}

type RevokeSessionRequest struct {
	SessionID int `json:"sid" validate:"required"`
}

type FindSessionResponse struct {
	ID      int       `json:"id"`
	Country string    `json:"country"`
	Created time.Time `json:"created"`
	Updated time.Time `json:"updated"`
	Revoked bool      `json:"revoked"`
	Count   int       `json:"-"`
}

type RevokeSessionResponse struct {
	ID      int       `json:"id"`
	Country string    `json:"country"`
	Created time.Time `json:"created"`
	Updated time.Time `json:"updated"`
	Revoked bool      `json:"revoked"`
}

// FindSession finds session based on parameters.
func FindSession(body *FindSessionRequest, session *claims.AuthClaims) (res []*FindSessionResponse, count int, err error) {
	res, err = database.Select[FindSessionResponse](
		findSessionTimeout,
		database.QuerySessionFind,
		session.UserID, body.ShowRevoked, body.Offset, body.Count,
	)

	// Check for error
	if err != nil {
		return
	}

	// Get total count
	count = 0
	if len(res) != 0 {
		count = res[0].Count
	}

	// Return result
	return
}

// RevokeSession revokes sessions from refreshing JWT-token.
func RevokeSession(body *RevokeSessionRequest, session *claims.AuthClaims) (res *RevokeSessionResponse, cookie *fiber.Cookie, err error) {
	res, err = database.SelectOne[RevokeSessionResponse](
		revokeSessionTimeout,
		database.QuerySessionRevoke,
		session.UserID, body.SessionID,
	)

	if err == pgx.ErrNoRows {
		err = ErrSessionNotFound
		return
	} else if err != nil {
		return
	}

	// Set expired cookie for refresh_token
	cookie = &fiber.Cookie{
		Name:     "refresh_token",
		Value:    "expired",
		Expires:  time.Time{},
		SameSite: "Strict",
		Path:     "/api/auth/refresh",
		HTTPOnly: true,
		Secure:   true,
	}

	// Return result
	return
}

File: frontend/src/api/session.ts
import { axios } from "../axios";

export interface Session {
  id: number;
  country: string;
  created: Date;
  updated: Date;
  revoked: boolean;
}

export interface FindSessionResponse {
  count: number;
  offset: number;
  sessions: Session[];
}

export interface RevokeSessionResponse {
  id: number;
  country: string;
  created: Date;
  updated: Date;
  revoked: boolean;
}

/**
 * Finds all sessions for authorized session.
 * Returned order is by latest.
 * Authorization required.
 */
export async function findSessions(req: {
  count: number;
  offset?: number;
}): Promise<FindSessionResponse> {
  const res = await axios.post("/api/session/find", req);
  return res.data;
}

/**
 * Revokes a session.
 * Returned order is by latest.
 * Authorization required.
 */
export async function revokeSession(req: {
  id: string;
}): Promise<RevokeSessionResponse> {
  const res = await axios.post("/api/session/revoke", req);
  return res.data;
}
